src/rdb/rdb_tx.c

```diff
@@ -756,6 +761,15 @@ rdb_tx_apply(struct rdb *db, uint64_t index, const void *buf, size_t len,
 					index, op.dto_opc, p - buf, n, rc);
 			break;
 		}
+
+		if ((op.dto_opc == RDB_TX_DELETE) ||
+		    (op.dto_opc == RDB_TX_DESTROY) ||
+		    (op.dto_opc == RDB_TX_DESTROY_ROOT))
+			crit = true;
```

Comment:
The tricky thing is, in theory, an RDB TX may include one of these "delete" operations, while also include operations that writing more data... :( In today's use cases, maybe we don't have such a TX (or do we?).

Reply:
I found at least one example: in pool destroy --force with open handles. MS sends POOL_EVICT RPC to the PS leader (with indication it is part of a destroy) who uses RDB_TX_DELETE for the pool handles, and RDB_TX_UPDATE for the number of handles, and another RDB_TX_UPDATE to mark the pool "not connectable". I can see this happen at the end of server/metadata.py metadatafillup, and in that test it is the first and only instance of a combination of critical/non-critical individual operations in rdb_tx_apply(). It seems the collection could be processed all as "critical" (since at least one of the items is), but would perhaps require in the worst case a first full additional pass through the operations in the buffer. I will look into this.

The above has not been shown in the test to fail (yet). However, in reproducing a test failure during the preceding container destroy loop, I noticed the "empty" log entry appended to accomplish the leadership verification in each rdb_tx_begin() was associated with another -DER_NOSPACE seen in rdb_raft_log_offer_single() [following the rdb_tx_apply and in the subsequent rdb_lc_update to persist the entry. So I will make sure these are treated as critical updates too.

Reply:
Ah, so this is but a feature! "If an operation is critical, then we want to the whole TX to commit", right? If so, it makes sense to me. We should have some control over how large these kind of TXs are.

Reply:
Yes this is what I have implemented in the latest patch. That's done in rdb_tx_append(), and the boolean is encoded at the beginning of the buffer that's interpreted in rdb_raft_log_offer_single().

If we put a limit on the size of the TX for which this tactic can be used, then I guess we still want to try to commit the TX, but without using VOS_OF_CRIT flag in the updates - i.e., in those mixed operation TXs, "undo" the critical property of the TX? And take our chances that it's not being run in a space pressure situation? Perhaps a D_WARN would make sense when the append code reverts the critical property on the TX? Also I would guess to put a limit of 8 operations (a goal being to not impact any known combined TXs that - the ones I have seen - are smaller).